IPython/core/compilerop.py

```diff
@@ -92,6 +92,9 @@ def __init__(self):
         # (otherwise we'd lose our tracebacks).
         linecache.checkcache = check_linecache_ipython
 
+        # Caching a dictionary { filename: execution_count } for nicely
+        # rendered tracebacks
+        self.filename_map = {}
```

Comment:
Your suggestion is taking the assumption that the filename is using the format `<ipython-input-{0}-{1}>` which is not the case in ipykernel, nor in xeus-python as they are not using the same `code_name` function. It's only the case when using the ipython shell directly.

Reply:
:/ I honestly wasn't aware this has impact on these projects.

That's a shame. It would be really unfortunate to make `CachingCompiler` responsible for remembering cell IDs. It just seems like a code smell to me.

I guess there isn't any better way currently to achieve this. I might go on a little crusade afterwards to put this dict somewhere else, just to ensure the stateless class remains stateless, but that's outside the scope of this PR.

BTW. do you happen to know where `xeus-python` depends on this?

Reply:
ipykernel and xeus-python provide their own `CachingCompiler` class to the shell, allowing them to use their own hash for the filename.

This is needed for the compatibility with the JupyterLab debugger. When stopping at a breakpoint, the filename used for AST compilation must refer to an actual file saved in a temporary folder for the debugger to retrieve the source.

> That's a shame. It would be really unfortunate to make CachingCompiler responsible for remembering cell IDs. It just seems like a code smell to me.

Why do you think it's a code smell? And why do you think it's important to have the `CachingCompiler` stateless?

I guess we could store the `{ filename: cell_id } ` mapping in the shell. But the caching compiler being already responsible for choosing the filename given the cell id, I find it to be the appropriate place.

Reply:
Also, the fact that the `CachingCompiler` actually caches data during the session does not sound crazy to me. It already manipulates the Python standard's `linecache` module which is also used for error tracebacks if I remember correctly.

Reply:
> Why do you think it's a code smell? And why do you think it's important to have the CachingCompiler stateless?

I guess it's 90% a matter of taste, but looking through the history of this code, the `Caching` was supposed to mean `manipulates the linecache`, not `holds data related to Code objects`. It just looks like this class is inching towards another responsibility, and it already has one that is quite difficult and technical. I would've liked it to stay stateless solely because its an interface to what looks like a hack, and knowing that it is stateless makes this hack that little bit easier to understand when reading the code.

This is in no way a deal breaker, but rather just me ensuring there's really no better way to do this. Seems like currently there isn't. If I'll find other places in the codebase where this kind of "find info about code cells" functionality is attached, I'll probably want to make a separate place for it.